File: llm_anthology/render_md.py
"""IR -> clean, portable Markdown (the 'keep' copy).

Content-faithful: Claude/ChatGPT/Gemini bodies are already Markdown, so text
passes through untouched. Hidden unicode is STRIPPED (this file is a portable
copy that may be re-fed to a model — no invisible injection payload survives).
The HTML renderer instead BADGES the same chars for forensic viewing.
"""
import json
import re

from llm_anthology import sanitize

_NEWLINES = re.compile(r"\s*[\r\n]+\s*")
# Characters that let a value break OUT of markdown link syntax
_URL_UNSAFE = {"(": "%28", ")": "%29", " ": "%20", "<": "%3C", ">": "%3E",
               '"': "%22", "[": "%5B", "]": "%5D", "\t": "%09", "\n": "", "\r": ""}


def _md_line(s):
    """A single-line field. Newlines are stripped so a value can never forge a
    turn header (`## Human`) inside the file we may re-feed to a model."""
    return _NEWLINES.sub(" ", _clean(s)).strip()


def _md_inline(s):
    """Inline text that sits inside markdown structure (e.g. link text)."""
    s = _md_line(s)
    return s.replace("\\", "\\\\").replace("[", "\\[").replace("]", "\\]")


def _md_code_span(s):
    """Content of a `...` span: no backticks, no newlines."""
    return _md_line(s).replace("`", "'")


def _md_url(u):
    """Link destination: percent-encode what would terminate the destination."""
    return "".join(_URL_UNSAFE.get(ch, ch) for ch in _clean(u))


def render_conversation_md(conv):
    out = ["# %s" % _md_line(conv.title), ""]
    meta = " · ".join(x for x in [
        conv.provider,
        ("account " + conv.account) if conv.account else "",
        ("created " + conv.created_at) if conv.created_at else "",
    ] if x)
    if meta:
        out += ["> " + meta, ""]
    for turn in conv.turns:
        who = "Human" if turn.role == "human" else "Assistant"
        hdr = "## %s" % who
        if turn.branch:
            hdr += "  _(branch %d/%d)_" % (turn.branch["index"], turn.branch["total"])
        out += [hdr, ""]
        for b in turn.blocks:
            out += [_block_md(b), ""]
        out += ["---", ""]
    return "\n".join(out).rstrip() + "\n"


def _clean(s):
    return sanitize.sanitize_for_copy(s or "")


def _block_md(b):
    if b.type == "text":
        body = _clean(b.text)
        parts = []
        for c in (b.citations or []):
            if not isinstance(c, dict):
                continue
            url = c.get("url") or ""
            title = _md_inline(c.get("title") or url or "source")
            # is_safe_url gates the SCHEME; the value must still be ENCODED, or a
            # ')' in the url (or a ']' in the title) forges a second, live link
            parts.append("[%s](%s)" % (title, _md_url(url)) if sanitize.is_safe_url(url)
                         else "%s (%s)" % (title, _md_inline(url)))
        if parts:
            body += "\n\nSources: " + " · ".join(parts)
        return body
    if b.type == "thinking":
        body = _clean(b.text)
        quoted = "\n".join("> " + ln for ln in body.splitlines()) if body else "> "
        return "> 🧠 **Thinking**\n>\n%s" % quoted
    if b.type == "tool_use":
        name = _md_code_span(b.data.get("name") or "tool")
        inp = b.data.get("input")
        inp_s = json.dumps(inp, ensure_ascii=False, indent=2) if inp is not None else ""
        disp = ("\n\n" + _clean(b.text)) if b.text else ""      # display_content
        return "**🔧 Tool call: `%s`**%s\n\n```json\n%s\n```" % (name, disp, _clean(inp_s))
    if b.type == "tool_result":
        name = _md_code_span(b.data.get("name") or "tool")
        content = b.data.get("content")
        cs = content if isinstance(content, str) else (
            json.dumps(content, ensure_ascii=False, indent=2) if content is not None else "")
        tag = "⚠️ error" if b.data.get("is_error") else "result"
        disp = ("\n\n" + _clean(b.text)) if b.text else ""      # display_content
        return "**↩️ Tool %s (`%s`)**%s\n\n```\n%s\n```" % (tag, name, disp, _clean(cs))
    if b.type == "attachment":
        d = b.data
        line = "📎 **%s** (%s, %s bytes)" % (
            _md_line(d.get("file_name") or "attachment"),
            _md_line(d.get("file_type") or "?"), d.get("file_size") or "?")
        ex = d.get("extracted_content")
        if ex:
            line += ("\n\n<details><summary>extracted content</summary>\n\n%s\n\n</details>" % _clean(ex))
        return line
    if b.type == "code":
        return "```%s\n%s\n```" % (_md_line(b.data.get("language") or ""), _clean(b.text))
    if b.type == "event":
        return "> ✨ **%s**" % _md_line(b.text or b.data.get("name") or "event")
    if b.type == "media":
        path = b.data.get("path") or ""
        # a REMOTE media url must not become a fetchable image in the portable copy
        # (it would beacon when this .md is rendered elsewhere) — mirror the HTML defang
        if "://" in path or path.startswith("//"):
            return "🖼 %s _(remote media — not embedded)_" % _md_inline(path)
        rel = path if "/" in path else ("../media/" + path)
        return "![%s](%s)" % (_md_inline(path or "media"), _md_url(rel))
    if b.type == "file":
        return "📎 %s _(no content in export)_" % _md_line(b.data.get("file_name") or "file")
    if b.type == "unknown":
        out = "_[unrendered %s block]_" % _md_line(str(b.data.get("orig_type") or "unknown"))
        raw = b.data.get("x_raw")
        try:
            body = json.dumps(raw, ensure_ascii=False, indent=2) if raw is not None else ""
        except (TypeError, ValueError):
            body = str(raw)
        if body:
            out += "\n\n```json\n%s\n```" % _clean(body)   # never hide a payload
        return out
    return _clean(b.text)

File: js/src/render_md.ts
/**
 * IR -> clean, portable Markdown (the "keep" copy). 1:1 with llm_anthology/render_md.py.
 *
 * Content-faithful: Claude/ChatGPT/Gemini bodies are already Markdown, so text
 * passes through untouched. Hidden unicode is STRIPPED (this file is a portable
 * copy that may be re-fed to a model — no invisible injection payload survives).
 * The HTML renderer instead BADGES the same chars for forensic viewing.
 *
 * Injection hardening that must not regress: newlines are stripped from every
 * single-line field so a value can never forge a `## Human` turn header, and link
 * titles/URLs are escaped and percent-encoded so a `)` or `]` in the data cannot
 * forge a second, live link.
 */
import type { Block, Conversation } from "./ir.js";
import { isSafeUrl, sanitizeForCopy } from "./sanitize.js";

const NEWLINES = /\s*[\r\n]+\s*/g;

/** Characters that let a value break OUT of markdown link syntax. */
const URL_UNSAFE: Record<string, string> = {
  "(": "%28",
  ")": "%29",
  " ": "%20",
  "<": "%3C",
  ">": "%3E",
  '"': "%22",
  "[": "%5B",
  "]": "%5D",
  "\t": "%09",
  "\n": "",
  "\r": "",
};

function clean(s: unknown): string {
  return sanitizeForCopy(s ?? "");
}

/** A single-line field. Newlines are stripped so a value can never forge a turn
 *  header inside the file we may re-feed to a model. */
function mdLine(s: unknown): string {
  return clean(s).replace(NEWLINES, " ").trim();
}

/** Inline text that sits inside markdown structure (e.g. link text). */
function mdInline(s: unknown): string {
  return mdLine(s).replace(/\\/g, "\\\\").replace(/\[/g, "\\[").replace(/\]/g, "\\]");
}

/** Content of a `...` span: no backticks, no newlines. */
function mdCodeSpan(s: unknown): string {
  return mdLine(s).replace(/`/g, "'");
}

/** Link destination: percent-encode what would terminate the destination. */
function mdUrl(u: unknown): string {
  return [...clean(u)].map((ch) => URL_UNSAFE[ch] ?? ch).join("");
}

function jsonDumps(value: unknown): string {
  return JSON.stringify(value, null, 2) ?? "";
}

export function renderConversationMd(conv: Conversation): string {
  const out: string[] = [`# ${mdLine(conv.title)}`, ""];
  const meta = [
    conv.provider,
    conv.account ? `account ${conv.account}` : "",
    conv.created_at ? `created ${conv.created_at}` : "",
  ]
    .filter((x) => x)
    .join(" · ");
  if (meta) out.push(`> ${meta}`, "");

  for (const t of conv.turns) {
    const who = t.role === "human" ? "Human" : "Assistant";
    let hdr = `## ${who}`;
    if (t.branch) hdr += `  _(branch ${t.branch.index}/${t.branch.total})_`;
    out.push(hdr, "");
    for (const b of t.blocks) out.push(blockMd(b), "");
    out.push("---", "");
  }
  return out.join("\n").replace(/\s+$/, "") + "\n";
}

function blockMd(b: Block): string {
  const d = b.data;
  const str = (k: string): string => (typeof d[k] === "string" ? (d[k] as string) : "");

  if (b.type === "text") {
    let body = clean(b.text);
    const parts: string[] = [];
    for (const c of b.citations) {
      if (c === null || typeof c !== "object") continue;
      const url = typeof c.url === "string" ? c.url : "";
      const title = mdInline(c.title || url || "source");
      // isSafeUrl gates the SCHEME; the value must still be ENCODED, or a ')' in
      // the url (or a ']' in the title) forges a second, live link
      parts.push(isSafeUrl(url) ? `[${title}](${mdUrl(url)})` : `${title} (${mdInline(url)})`);
    }
    if (parts.length) body += "\n\nSources: " + parts.join(" · ");
    return body;
  }

  if (b.type === "thinking") {
    const body = clean(b.text);
    const quoted = body
      ? body.split("\n").map((ln) => "> " + ln).join("\n")
      : "> ";
    return `> 🧠 **Thinking**\n>\n${quoted}`;
  }

  if (b.type === "tool_use") {
    const name = mdCodeSpan(str("name") || "tool");
    const inp = d["input"];
    const inpS = inp !== undefined && inp !== null ? jsonDumps(inp) : "";
    const disp = b.text ? "\n\n" + clean(b.text) : ""; // display_content
    return `**🔧 Tool call: \`${name}\`**${disp}\n\n\`\`\`json\n${clean(inpS)}\n\`\`\``;
  }

  if (b.type === "tool_result") {
    const name = mdCodeSpan(str("name") || "tool");
    const content = d["content"];
    const cs =
      typeof content === "string"
        ? content
        : content !== undefined && content !== null
          ? jsonDumps(content)
          : "";
    const tag = d["is_error"] ? "⚠️ error" : "result";
    const disp = b.text ? "\n\n" + clean(b.text) : "";
    return `**↩️ Tool ${tag} (\`${name}\`)**${disp}\n\n\`\`\`\n${clean(cs)}\n\`\`\``;
  }

  if (b.type === "attachment") {
    let line = `📎 **${mdLine(str("file_name") || "attachment")}** (${
      mdLine(str("file_type") || "?")
    }, ${d["file_size"] ?? "?"} bytes)`;
    const ex = d["extracted_content"];
    if (ex) {
      line += `\n\n<details><summary>extracted content</summary>\n\n${clean(ex)}\n\n</details>`;
    }
    return line;
  }

  if (b.type === "code") {
    return `\`\`\`${mdLine(str("language"))}\n${clean(b.text)}\n\`\`\``;
  }

  if (b.type === "event") {
    return `> ✨ **${mdLine(b.text || str("name") || "event")}**`;
  }

  if (b.type === "media") {
    const path = str("path");
    // a REMOTE media url must not become a fetchable image in the portable copy
    // (it would beacon when this .md is rendered elsewhere) — mirror the HTML defang
    if (path.includes("://") || path.startsWith("//")) {
      return `🖼 ${mdInline(path)} _(remote media — not embedded)_`;
    }
    const rel = path.includes("/") ? path : "../media/" + path;
    return `![${mdInline(path || "media")}](${mdUrl(rel)})`;
  }

  if (b.type === "file") {
    return `📎 ${mdLine(str("file_name") || "file")} _(no content in export)_`;
  }

  if (b.type === "unknown") {
    let out = `_[unrendered ${mdLine(String(d["orig_type"] ?? "unknown"))} block]_`;
    const raw = d["x_raw"];
    let body = "";
    try {
      body = raw !== undefined && raw !== null ? jsonDumps(raw) : "";
    } catch {
      body = String(raw);
    }
    if (body) out += `\n\n\`\`\`json\n${clean(body)}\n\`\`\``; // never hide a payload
    return out;
  }

  return clean(b.text);
}
